internal/mailer/mailme.go

```diff
@@ -0,0 +1,193 @@
+}
+
+// Mail sends a templated mail. It will try to load the template from a URL, and
+// otherwise fall back to the default
+func (m *MailmeMailer) Mail(to, subjectTemplate, templateURL, defaultTemplate string, templateData map[string]interface{}) error {
+	if m.FuncMap == nil {
+		m.FuncMap = map[string]interface{}{}
+	}
+	if m.cache == nil {
+		m.cache = &TemplateCache{
+			templates: map[string]*MailTemplate{},
+			funcMap:   m.FuncMap,
+			logger:    m.Logger,
+		}
+	}
+
+	tmp, err := template.New("Subject").Funcs(template.FuncMap(m.FuncMap)).Parse(subjectTemplate)
+	if err != nil {
+		return err
+	}
+
+	subject := &bytes.Buffer{}
+	err = tmp.Execute(subject, templateData)
+	if err != nil {
+		return err
+	}
+	body, err := m.MailBody(templateURL, defaultTemplate, templateData)
+	if err != nil {
+		return err
+	}
+
+	mail := gomail.NewMessage()
+	mail.SetHeader("From", m.From)
+	mail.SetHeader("To", to)
+	mail.SetHeader("Subject", subject.String())
+	mail.SetBody("text/html", body)
+
+	dial := gomail.NewDialer(m.Host, m.Port, m.User, m.Pass)
+	if m.LocalName != "" {
+		dial.LocalName = m.LocalName
+	}
+	return dial.DialAndSend(mail)
+
+}
+
+type MailTemplate struct {
+	tmp       *template.Template
+	expiresAt time.Time
+}
+
+type TemplateCache struct {
+	templates map[string]*MailTemplate
+	mutex     sync.Mutex
+	funcMap   template.FuncMap
+	logger    logrus.FieldLogger
+}
+
+func (t *TemplateCache) Get(url string) (*template.Template, error) {
+	cached, ok := t.templates[url]
+	if ok && (cached.expiresAt.Before(time.Now())) {
+		return cached.tmp, nil
+	}
+	data, err := t.fetchTemplate(url, TemplateRetries)
+	if err != nil {
+		return nil, err
+	}
+	return t.Set(url, data, TemplateExpiration)
+}
+
+func (t *TemplateCache) Set(key, value string, expirationTime time.Duration) (*template.Template, error) {
+	parsed, err := template.New(key).Funcs(t.funcMap).Parse(value)
+	if err != nil {
+		return nil, err
+	}
+
+	cached := &MailTemplate{
+		tmp:       parsed,
+		expiresAt: time.Now().Add(expirationTime),
+	}
+	t.mutex.Lock()
+	t.templates[key] = cached
+	t.mutex.Unlock()
+	return parsed, nil
+}
+
+func (t *TemplateCache) fetchTemplate(url string, triesLeft int) (string, error) {
+	client := &http.Client{
+		Timeout: 10 * time.Second,
+	}
+
+	resp, err := client.Get(url)
+	if err != nil && triesLeft > 0 {
+		return t.fetchTemplate(url, triesLeft-1)
+	}
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode == 200 { // OK
+		bodyBytes, err := io.ReadAll(resp.Body)
+		if err != nil && triesLeft > 0 {
+			return t.fetchTemplate(url, triesLeft-1)
+		}
+		if err != nil {
+			return "", err
+		}
+		return string(bodyBytes), err
+	}
+	if triesLeft > 0 {
+		return t.fetchTemplate(url, triesLeft-1)
+	}
+	return "", errors.New("mailer: unable to fetch mail template")
+}
+
+func (m *MailmeMailer) MailBody(url string, defaultTemplate string, data map[string]interface{}) (string, error) {
+	if m.FuncMap == nil {
```

Comment:
We want to set this at time of New() too. I can do a quick patch to it after it's in if you want to fix the races.